src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1315 @@
+            await self.update_netbios_data(old, new)
+        except Exception as e:
+            raise ValidationError('netbiosname', str(e))
+
+        new = await self.ad_compress(new)
+
+        if not new["bindpw"] and not new["kerberos_principal"]:
+            raise ValidationError("activedirectory_update.bindname", "Bind credentials or kerberos keytab are required to join an AD domain.")
+
+        if data['enable'] and not old['enable']:
+            try:
+                await self.middleware.run_in_thread(self.validate_credentials, new)
+            except Exception as e:
+                verrors.add("activedirectory_update.bindpw", f"Failed to validate bind credentials: {e}")
+            try:
+                await self.middleware.run_in_thread(self.validate_domain, new)
+            except Exception as e:
+                verrors.add("activedirectory_update", f"Failed to validate domain configuration: {e}")
+
+        if verrors:
+            raise verrors
+
+        await self.middleware.call(
+            'datastore.update',
+            'directoryservice.activedirectory',
+            old['id'],
+            new,
+            {'prefix': 'ad_'}
+        )
+
+        start = False
+        stop = False
+
+        if old['idmap_backend'] != new['idmap_backend']:
+            idmap = await self.middleware.call('idmap.domaintobackend.query', [('domain', '=', 'DS_TYPE_ACTIVEDIRECTORY')])
+            await self.middleware.call('idmap.domaintobackend.update', idmap[0]['id'], {'idmap_backend': new['idmap_backend']})
+
+        if not old['enable']:
+            if new['enable']:
+                start = True
+        else:
+            if not new['enable']:
+                stop = True
+
+        if stop:
+            await self.stop()
+        if start:
+            await self.start()
+
+        if not stop and not start and new['enable']:
+            await self.middleware.call('service.restart', 'cifs')
+
+        return await self.config()
+
+    @private
+    async def _set_state(self, state):
+        await self.middleware.call('cache.put', 'AD_State', state.name)
+
+    @accepts()
+    async def get_state(self):
+        """
+        Check the state of the AD Directory Service.
+        See DSStatus for definitions of return values.
+        :DISABLED: Service is not enabled.
+        If for some reason, the cache entry indicating Directory Service state
+        does not exist, re-run a status check to generate a key, then return it.
+        """
+        ad = await self.config()
+        if not ad['enable']:
+            return 'DISABLED'
+        else:
+            try:
+                return (await self.middleware.call('cache.get', 'AD_State'))
+            except KeyError:
+                try:
+                    await self.started()
+                except Exception:
+                    pass
+
+            return (await self.middleware.call('cache.get', 'AD_State'))
+
+    @private
+    async def start(self):
```

Comment:
How long can start take?

I am thinking if perhaps we should let it be a job so we can report progress?

Reply:
Let me elaborate a bit more, we should only consider that if joining AD can take a lot of time so the user can have an idea what its doing. Either way, not urgent and can be addressed in another PR. Just wanted to keep that in mind.

Reply:
That's a good thought. It doesn't take as long as it used to (the largest culprit before was filling the AD cache and initializing everything through freenasldap.py). I think we're at less than 10 seconds assuming a healthy AD environment. Throwing it in a job might be a good idea for the future if this proves to be problematic.

Reply:
Fair enough, ~10 seconds is fast enough